Add zircote/swagger-php attributes support

sonrac opened the pull request:
Add zircote/swagger-php attributes support

sonrac committed:
add swagger-php attributes support and supports symfony with version >= 6.0

sonrac committed:
Create php.yml

sonrac committed:
remove unusages workflow

sonrac committed:
Merge remote-tracking branch 'origin/master' into add_swagger_attributes_support

sonrac committed:
remove double slashes

sonrac committed:
fix tests

Tobion commented on tests/Annotations/OpenApiRouteLoaderAnnotationsTest.php:
```diff
@@ -0,0 +1,233 @@
+{
+    private const FIXTURES_ROUTE_NAME_PREFIX = 'tobion_openapisymfonyrouting_tests_fixtures_annotations_';
+
+    public function testBasic(): void
```
Instead of re-implementing the tests for attributes and annotations, I'd suggest to keep the current tests of OpenApiRouteLoaderTest but make it abstract. Then extend this base test class from OpenApiRouteLoaderAnnotationsTest and OpenApiRouteLoaderAttributesTest. Both tests cases should be exactly the same. Only the fixtures base path changes.

sonrac replied:
It was some problems with swagger ^4.0 

Swagger by default generates hash for operationId by default 
see https://github.com/zircote/swagger-php/blob/master/src/Processors/OperationId.php#L23

Format for ^4.0 does not assert with the previous version, that's why I removed operationId generation checks. 

I think that abstract tests it bad idea, tests workflow will be changed for some environments and it will open some problems with changing the tests workflow if you have abstract tests for both environments.

Tobion replied:
The operatinId processor was removed before and should also be removed in the attribute version. Analysis::processors does not exist anymore in 4.x it seems. But we can call Generator::getProcessors and Generator::setProcessors to remove the OpererationId processor: https://github.com/zircote/swagger-php/blob/3e41d604d6d1496043be1de9f688db410177f384/src/Generator.php#L183-L217
This should also work in 3.x to make it the same for attributes and annoatations.

sonrac replied:
@Tobion  OperationID Processor removed

Tobion commented:
Thank you for working on this. I guess this will fix #12

Tobion commented on phpunit.xml:
```diff
@@ -9,7 +9,7 @@
     <testsuites>
         <testsuite name="Test Suite">
-            <directory>tests/</directory>
+            <directory>tests/Annotations</directory>
```
why this? should not be needed.

sonrac replied:
tests for attributes do not work with PHP < 8.1

Maybe it may have the meaning to create a new version only for PHP 8.1? Swagger PHP Attributes does not work with PHP version < 8.1 properly.  

My current version tries working with all PHP versions but in CI we need to remember these restrictions. But as I say you will have many problems with version PHP 8.0. 

Some does not support features (nested attributes), for example, some dirty hack from this what I saw in PHP 8.0

```
define("CONST_OBJ_NAME", new OA\Schema(type: string))

#OA\Parameter(name: 'some_attr', schema: CONST_OBJ_NAME)
```

Tobion replied:
Yeah attributes will only be supported with php 8.1. Please add `@requires PHP >= 8.1` as annotation to the attributes test class. This way the tests are skipped in lower php versions.

sonrac replied:
@Tobion  done

Tobion commented on .github/workflows/ci.yml:
```diff
@@ -76,4 +76,4 @@ jobs:
             -   name: Run tests
                 continue-on-error: true
-                run: vendor/bin/simple-phpunit
+                run: vendor/bin/simple-phpunit tests
```
should not be needed. that  is what phpunit.xml should configure already

sonrac committed:
add php attributes support

https://github.com/Tobion/OpenAPI-Symfony-Routing/issues/12 add path parameter enum

Also need to add tests for new features like OA\PathParameter and enums, see example zircote/swagger-php#1060

Tobion commented on tests/Attributes/Fixtures/Basic/Controller.php:
```diff
@@ -0,0 +1,17 @@
+
+namespace Tobion\OpenApiSymfonyRouting\Tests\Attributes\Fixtures\Basic;
+
+use OpenApi\Attributes as OAT;
```
please use only `OA` as name to be consistent with the annotations and swagger-php doc

Tobion commented on .gitignore:
```diff
@@ -1,3 +1,7 @@
 vendor/
+docker-compose.override.yml
+/composer
+/.idea
```
those things are better suited to add to your own global gitignore

Tobion commented on composer.json:
```diff
@@ -16,7 +16,7 @@
         "symfony/framework-bundle": "^4.4|^5.0|^6.0",
         "symfony/routing": "^4.4|^5.0|^6.0",
-        "zircote/swagger-php": "^3.0.3"
+        "zircote/swagger-php": "^3.0.3|^4.0"
```
"^3.0.3|^4.1" because 4.0 is not compatible as the attributes namespace changed

mario-fehr commented:
@sonrac @Tobion any news on this?
